language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/toml/ballerinatoml/completion/BallerinaTomlCompletionUtil.java

```diff
@@ -0,0 +1,98 @@
+/*
+ *  Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ *  WSO2 Inc. licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+package org.ballerinalang.langserver.toml.ballerinatoml.completion;
+
+import io.ballerina.toml.syntax.tree.DocumentNode;
+import io.ballerina.toml.syntax.tree.Node;
+import io.ballerina.toml.syntax.tree.NonTerminalNode;
+import io.ballerina.toml.syntax.tree.TableArrayNode;
+import io.ballerina.toml.syntax.tree.TableNode;
+import org.ballerinalang.langserver.commons.LanguageServerContext;
+import org.ballerinalang.langserver.commons.toml.TomlCompletionContext;
+import org.ballerinalang.langserver.commons.toml.common.completion.TomlCompletionUtil;
+import org.ballerinalang.langserver.commons.toml.visitor.TomlNode;
+import org.eclipse.lsp4j.CompletionItem;
+import org.eclipse.lsp4j.jsonrpc.messages.Either;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Responsible for providing completion items based on the context.
+ *
+ * @since 2.0.0
+ */
+public class BallerinaTomlCompletionUtil {
+
+    public static List<CompletionItem> getCompletionItems(TomlCompletionContext ctx,
```

Comment:
This implementation is the behavior for generic toml file with a schema right? Can we move this to an abstract class(default impl) or move to util so it can be reused by other implementations?

Reply:
Common utilities are moved to TomlCompletionUtil class. Extension developers can utilize those methods as  I have used in here.  Since this is a static utility class I thought it is not suitable to realize an abstract class. WDYT?

Reply:
Does this have to be static? If you take the Cloud.toml scenario, as of now, we generate the completions based on the schema. As far as i understood, we get all the possible supported completions and then remove the existing entries based on the context. If so, the getCompletionItems logic would be the exact same thing for us. I think this should be the default case for any toml. Of course anyone should be able to override or modify that behavior. I think it would be great if we can avoid that code duplication.

Reply:
@xlight05 yes, there is a code duplication. will address this.